README.md
# MCP Cache

[![CircleCI](https://circleci.com/gh/QuickenLoans/mcp-cache.svg?style=shield)](https://circleci.com/gh/QuickenLoans/mcp-cache)
[![Latest Stable Version](https://poser.pugx.org/ql/mcp-cache/version)](https://packagist.org/packages/ql/mcp-cache)
[![License](https://poser.pugx.org/ql/mcp-cache/license)](https://packagist.org/packages/ql/mcp-cache)

Simple caching standard for Quicken Loans PHP projects.

We prefer a simpler cache protocol than the one provided by [PSR-6 Caching Interface](http://www.php-fig.org/psr/psr-6/).
This package has many similarities with the draft [PSR-16 Simple Cache](https://github.com/php-fig/fig-standards/blob/master/proposed/simplecache.md).
If this PSR is accepted, we will update this package to ensure compatibility.

## Installation

```
composer require ql/mcp-cache ~3.0
```

## Table of Contents

- [Usage](#usage)
    - [CachingTrait](#cachingtrait)
- [Implementations](#implementations)
    - [MemoryCache](#memorycache)
    - [PredisCache](#prediscache)
    - [APCCache](#apccache)
    - [MemcachedCache](#memcachedcache)
    - [MemcacheCache](#memcachecache)
- [Stampede Protection](#stampede-protection)


## Usage

The `CacheInterface` has only two methods - `get` and `set`.

### Get

```php
public function get($key);
```

Get takes any string as a key and does a lookup of the data.

##### Please Note:
- Missing data will return `null`. There is no difference between a cache hit of a `null` value and a cache miss.
- It is not necessary to serialize objects before storage. Serialization will be optimized by each cache implementation.

### Set

```php
public function set($key, $value, $ttl = 0);
```

Set data in the cache. A boolean will be returned to indicate whether the data was saved.

##### Please Note:
- Resources cannot be cached.
- TTL is the time in seconds the data should live until expired. A time to live of `0` will never expire the data.

### CachingTrait

The [CachingTrait](src/CachingTrait.php) is provided to make adding optional caching to your classes easy.

`QL\MCP\Cache\CachingTrait` adds the following private methods to your trait consumer:

```php
/**
 * @return CacheInterface|null
 */
private function cache();

/**
 * @param string $key
 * @return mixed|null
 */
private function getFromCache($key);

/**
 * @param string $key
 * @param mixed $value
 * @param mixed $ttl
 * @return null
 */
private function setToCache($key, $value, $ttl = 0);
```

If you want to access the cache directly, use `$this->cache()`. However, it is usually not necessary. The `get` and `set` methods will fail gracefully and return `null` if no cacher was set.

The following public methods are added as well:
```php
/**
 * @param CacheInterface $cache
 * @return null
 */
public function setCache(CacheInterface $cache);

/**
 * @param int $ttl
 * @return null
 */
public function setCacheTTL($ttl);
```

Here is an example of this in action:

```php
use QL\MCP\Cache\CachingTrait;

class MyRepo
{
    use CachingTrait;

    public function load()
    {
        if ($cached = $this->getFromCache('cachekey')) {
            return $cached;
        }

        // get data from service
        $data = $this->callService();

        $this->setToCache('cachekey', $data);
        return $data;
    }
}
```

Setup example for **Symfony DI**:
```
services:
    cache:
        class: 'QL\MCP\Cache\PredisCache'
        arguments: ['@predis']
    repo:
        class: 'MyRepo'
        calls:
            - ['setCache', ['@cache']]
            - ['setCacheTTL', [3600]]
```
Setup example for **PHP**:
```php
$cacher = new PredisCache($predis);

$repo = new MyRepo;
$repo->setCache($cacher);
$repo->setCacheTTL(3600);
```

#### Why use this?

- Reduce boilerplate
- Make caching easy

It allows you to enable or disable caching with minimal boilerplate. All of the caching actions within your class will fail gracefully if the cache was never set.

Please note that it is possible to set a *global* ttl for your class. If a ttl is provided at the time of setting data, it will be used. If not provided, the global ttl will be used. If the global ttl is never set, no ttl will be used.

## Implementations

- [MemoryCache](#memorycache)
- [PredisCache](#prediscache)
- [APCCache](#apccache)
- [MemcachedCache](#memcachedcache)
- [MemcacheCache](#memcachecache)

### MemoryCache

The `MemoryCache` is a very basic cache for caching data that only lives through the lifetime
of the request. This cache ignores `ttl`.

```php
use QL\MCP\Cache\MemoryCache;

$cache = new MemoryCache

// Store data
$cache->set('key', $data);

// Store data with expiration of 10 minutes  - note that ttl is ignored for this cacher
$cache->set('key', $data, 600);
```

### PredisCache

This cache will store data in the redis using [predis](https://github.com/nrk/predis).

An optional suffix may be provided to salt the cache keys. This can be used to invalidate the entire cache
between code pushes or other configuration changes.

Setting a key to `null` will delete the key rather than setting the value to `null`.
It is not possible to store a `null` value with the predis cacher.

```php
use Predis\Client;
use QL\MCP\Cache\PredisCache;

$predis = new Client;
$suffix = '6038aa7'; // optional
$cache = new PredisCache($predis, $suffix);

// Store data
$cache->set('key', $data);

// Store data with expiration of 10 minutes
$cache->set('key', $data, 600);
```

### APCCache

This cache will store items in the APC user cache space. Optionally, a maximum TTL may be set by calling the 
`APCCache::setMaximumTtl($ttl)` method.

```php
use QL\MCP\Cache\APCCache;
use QL\MCP\Common\Time\Clock;

$cache = new APCCache(new Clock());

// Optionally, set the maximum TTL to 10 minutes
$cache->setMaximumTtl(600);

// Store data
$cache->set('key', $data);

// Store data with an expiration of 10 minutes
$cache->set('key', $data, 600);

// Retrieve data
$data = $cache->get('key');
```

### MemcachedCache

This cache will store data in memcached using [pecl-memcached](https://github.com/php-memcached-dev/php-memcached).
It is also compatible with [AWS ElastiCache Cluster Client](https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-php),
a drop-in replacement for `pecl-memcache` with support for autodiscovery.

See [PHP.NET Memcached Book](http://php.net/manual/en/book.memcached.php)

### MemcacheCache

This cache will store data in memcached using [pecl-memcache](https://github.com/php-memcached-dev/php-memcache).

See [PHP.NET Memcache Book](http://php.net/manual/en/book.memcache.php)

## Stampede Protection

In the case of high load services, when the cache expires or is flushed many requests attempting to regenerate the
cache can cause a dog-piling effect on dependent ssystems, especially if the cost of regenerating the cached data is high.
This is typically a concern under heavy load, when cached data is shared across many requests.

Several methods exist for preventing this, for MCP Cache we implement **Probabilistic early expiration**.

With this approach, when the remaining TTL gets close to expiring, the application has a higher and higher random
chance of returning a **cache miss** from `get`. In this scenario, a very small portion of users will attempt to regenerate
the cache, rather than every request.

## Caching implementations with Stampede Protection

Note: by default stampede protection is **disabled**.

- [APCCache](#apccache)

### Code example

```php
use QL\MCP\Cache\APCCache;

$cache = new APCCache;
$cache->enableStampedeProtection();

// Customize beta and delta (not recommended).
$cache->setPrecomputeBeta(5);
$cache->setPrecomputeDelta(10);

// use cache as normal
$cache->set('test', 'value', 60);
```

### Example Cache Stampede Protection

Beta = `5`  
Delta = `10` % of TTL

Set a value with a ttl of 60 seconds and run 1000 tests.

TTL left   | percentile | early expires   | percent
---------- | ---------- | --------------- | -------
25s        | 60%        |   0 out of 1000 | 0%
20s        | 66%        |  10 out of 1000 | 1%
15s        | 75%        |  40 out of 1000 | 4%
6s         | 90%        | 300 out of 1000 | 30%
3s         | 95%        | 500 out of 1000 | 50%

References:
- [https://en.wikipedia.org/wiki/Cache_stampede#Probabilistic_early_expiration](https://en.wikipedia.org/wiki/Cache_stampede#Probabilistic_early_expiration)
- [http://www.vldb.org/pvldb/vol8/p886-vattani.pdf](http://www.vldb.org/pvldb/vol8/p886-vattani.pdf)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. See [keepachangelog.com](http://keepachangelog.com) for reference.

This package follows [semver](http://semver.org/) versioning.

## [3.1.0] - 2019-04-13

### Changed
- `ql/mcp-common` >= 2.0 is now required.
- PHP >= 7.1 is now required.

### Removed
- `MemcacheCache` has been removed. `ext-memcache` is not PHP 7 compatible and this library now requires PHP 7.1 or higher.

## [3.0.0] - 2016-11-21

### Added
- Added `CacheInterface::VERSION` to get the current version of the Cache package.
- **[BC WARNING]** Append cache version to cache keys automatically.
    - When the cache package is updated, old data will be automatically ignored
      (Ensure your caches are always in [LRU mode](https://en.wikipedia.org/wiki/Cache_algorithms#LRU)!).
      This will make it easier for consumers to handle cached data that is not backwards compatible.

### Changed
- Changed namespace from `MCP\Cache\*` to `QL\MCP\Cache\*`.
- PHP `5.6+` or `7.0+` is now required.
- Switched dependency from `ql/mcp-core` to `ql/mcp-common`.
- Fixed `APCCache::clear()`.

### Removed

- Removed `SkeletorSessionCache`.

## [2.5.1] - 2015-10-13

### Changed
- Cache keys can now be salted with a suffix for **MemcachedCache**.
- A maximum TTL can now be set for **MemcachedCache**.
- Error responses from **MemcachedCache** can now be optionally logged.
    - `pecl-memcached` returns error codes when failures occur, which **MemcachedCache** silently consumes. Attach
      a logger to log these messages as warnings.

## [2.5.0] - 2015-10-08

### Added
- Add **MemcacheCache** for caching using `pecl-memcache`.
- Add **MemcachedCache** for caching using `pecl-memcached`.

## [2.4.1] - 2015-09-21

### Changed
- **APCCache** now uses `apcu_*` methods instead of `apc_*`.

## [2.4.0] - 2015-09-01

### Added
- Add optional cache stampede protection to **APCCache**. This is disabled by default.

### Changed
- Cache keys can now be salted with a suffix for **APCCache**.
- Attempts to cache a resource will now throw `MCP\Cache\Exception` instead of `InvalidArgumentException`.

## [2.3.1] - 2015-05-01

### Changed
- A maximum TTL can now be set for **APCCache**.

## [2.3.0] - 2015-05-01

### Added
- Add **APCCache** for caching to APC (APCU recommended).

## [2.2.0] - 2014-11-12

### Added
- Add `MCP\Cache\CachingTrait` for enabling quick cache integration on repositories, services, or other code.


src/MemcachedCache.php
<?php
/**
 * @copyright (c) 2016 Quicken Loans Inc.
 *
 * For full license information, please view the LICENSE distributed with this source code.
 */

namespace QL\MCP\Cache;

use Memcached;
use Psr\Log\LoggerInterface;
use QL\MCP\Cache\Utility\KeySaltingTrait;
use QL\MCP\Cache\Utility\MaximumTTLTrait;

/**
 * PHP.NET Docs:
 *
 * @see http://php.net/manual/en/book.memcached.php
 *
 * GitHub Source:
 *
 * @see https://github.com/php-memcached-dev/php-memcached
 * @see https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-php
 *
 * ElastiCache Docs:
 *
 * @see http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/Appendix.PHPAutoDiscoverySetup.html
 * @see http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.Using.html
 */
class MemcachedCache implements CacheInterface
{
    use KeySaltingTrait;
    use MaximumTTLTrait;

    const ERR_GET = 'Memcached Error : GET : %s';
    const ERR_SET = 'Memcached Error : SET : %s';

    const UNKNOWN_CODE = '?';

    /**
     * Properties read and used by the salting trait.
     *
     * @var string
     */
    const PREFIX = 'mcp-cache-' . CacheInterface::VERSION;
    const DELIMITER = ':';

    /**
     * @var Memcached
     */
    private $cache;

    /**
     * @var LoggerInterface|null
     */
    private $logger;

    /**
     * @var string|null
     */
    private $suffix;

    /**
     * To generate list:
     *
     * ```php
     * $results = [];
     * array_walk((new ReflectionClass('Memcached'))->getConstants(), function($v, $k) use (&$results) {
     *     if (substr($k, 0, 4) === 'RES_') $results[$k] = $v;
     * });
     * var_dump($results);
     * ```
     *
     * @var array
     */
    private $resultCodes = [
        'RES_SUCCESS' => 0,
        'RES_FAILURE' => 1,
        'RES_HOST_LOOKUP_FAILURE' => 2,
        'RES_UNKNOWN_READ_FAILURE' => 7,
        'RES_PROTOCOL_ERROR' => 8,
        'RES_CLIENT_ERROR' => 9,
        'RES_SERVER_ERROR' => 10,
        'RES_WRITE_FAILURE' => 5,
        'RES_DATA_EXISTS' => 12,
        'RES_NOTSTORED' => 14,
        'RES_NOTFOUND' => 16,
        'RES_PARTIAL_READ' => 18,
        'RES_SOME_ERRORS' => 19,
        'RES_NO_SERVERS' => 20,
        'RES_END' => 21,
        'RES_ERRNO' => 26,
        'RES_BUFFERED' => 32,
        'RES_TIMEOUT' => 31,
        'RES_BAD_KEY_PROVIDED' => 33,
        'RES_STORED' => 15,
        'RES_DELETED' => 22,
        'RES_STAT' => 24,
        'RES_ITEM' => 25,
        'RES_NOT_SUPPORTED' => 28,
        'RES_FETCH_NOTFINISHED' => 30,
        'RES_SERVER_MARKED_DEAD' => 35,
        'RES_UNKNOWN_STAT_KEY' => 36,
        'RES_INVALID_HOST_PROTOCOL' => 34,
        'RES_MEMORY_ALLOCATION_FAILURE' => 17,
        'RES_CONNECTION_SOCKET_CREATE_FAILURE' => 11,
        'RES_E2BIG' => 37,
        'RES_KEY_TOO_BIG' => 39,
        'RES_SERVER_TEMPORARILY_DISABLED' => 47,
        'RES_SERVER_MEMORY_ALLOCATION_FAILURE' => 48,
        'RES_AUTH_PROBLEM' => 40,
        'RES_AUTH_FAILURE' => 41,
        'RES_AUTH_CONTINUE' => 42,
        'RES_PAYLOAD_FAILURE' => -1001,
    ];

    /**
     * An optional suffix can be provided which will be appended to the key
     * used to store and retrieve data. This can be used to throw away cached
     * data with a code push or other configuration change.
     *
     * Optionally, set a logger to enable logging of non-successful memcached responses.
     * - Error, If no servers in memcached pool
     * - Warning, all other result codes
     *
     * @param Memcached $cache
     * @param string|null $suffix
     * @param LoggerInterface|null $logger
     */
    public function __construct(Memcached $cache, $suffix = null, LoggerInterface $logger = null)
    {
        $this->cache = $cache;
        $this->suffix = $suffix;
        $this->logger = $logger;

        // Flip the result codes dict for efficiency.
        $this->resultCodes = array_flip($this->resultCodes);
    }

    /**
     * @see http://docs.php.net/manual/en/memcached.get.php
     *
     * {@inheritdoc}
     */
    public function get($key)
    {
        $key = $this->salted($key, $this->suffix);

        $cached = $this->cache->get($key);
        $code = $this->cache->getResultCode();

        if ($code === Memcached::RES_SUCCESS) {
            return $cached;

        } elseif ($code !== Memcached::RES_NOTFOUND) {
            $this->sendAlert('get', $key, $code);
        }

        return null;
    }

    /**
     * @see http://docs.php.net/manual/en/memcached.set.php
     *
     * {@inheritdoc}
     */
    public function set($key, $value, $ttl = 0)
    {
        $key = $this->salted($key, $this->suffix);

        // handle deletions
        if ($value === null) {
            $this->cache->delete($key);
            return true;
        }

        // Resolve the TTL to use
        $ttl = $this->determineTTL($ttl);

        $this->cache->set($key, $value, $ttl);
        $code = $this->cache->getResultCode();

        if ($code === Memcached::RES_SUCCESS) {
            return true;
        }

        $this->sendAlert('set', $key, $code);
        return false;
    }

    /**
     * @param int $code
     *
     * @return string|int
     */
    private function getHumanResultCode($code)
    {
        if (isset($this->resultCodes[$code])) {
            return $this->resultCodes[$code];
        }

        if (is_scalar($code)) {
            return $code;
        }

        return self::UNKNOWN_CODE;
    }

    /**
     * @param string $type
     * @param string $key
     * @param mixed $code
     *
     * @return void
     */
    private function sendAlert($type, $key, $code)
    {
        if (!$this->logger) {
            return;
        }

        $template = ($type === 'get') ? self::ERR_GET : self::ERR_SET;
        $priority = ($code === Memcached::RES_NO_SERVERS) ? 'error' : 'warning';
        $code = $this->getHumanResultCode($code);

        $msg = sprintf($template, $code);

        $this->logger->$priority($msg, [
            'cacheKey' => $key,
            'memcacheError' => $code,
        ]);
    }
}
